scripts/configure

```diff
@@ -0,0 +1,337 @@
+# configure will override the toolchain for the GN build using a 'custom'
+# toolchain assembled from the usual environment variables (CC, CXX, AR, CFLAGS,
+# CXXFLAGS, ...).
+
+function usage() {
+    info "Usage: $0 [OPTIONS] [--project=... [PROJECT OPTIONS]]"
+    info "Options:"
+    info "  --build-env-dir=DIR     Directory to create (host) build environment in"
+    info "  --project=DIR           Sub-directory to build, e.g. examples/lighting-app/linux"
+    exit 0
+}
+
+function main() { # ...
+    set -o pipefail
+    CHIP_ROOT=$(cd "$(dirname "$0")/.." && pwd)
+    if [[ "$PWD" == "$CHIP_ROOT" ]]; then
+        BUILD_DIR="out/configured"
+        BUILD_ROOT="${CHIP_ROOT}/${BUILD_DIR}"
+        BUILD_ENV_DIR=".venv"
+        info "Configuring in-tree, will build in $BUILD_DIR using environment $BUILD_ENV_DIR"
+    else
+        BUILD_DIR="."
+        BUILD_ROOT="$PWD"
+        BUILD_ENV_DIR="build-env"
+    fi
+    PROJECT=
+
+    # Parse main options but leave project options in $@
+    while [[ $# -gt 0 && -z "$PROJECT" ]]; do
+        case "$1" in
+        --help) usage ;;
+        --build-env-dir=*) BUILD_ENV_DIR="${1#*=}" ;;
+        --project=*) PROJECT="${1#*=}" ;;
+        *) fail "Invalid argument: '$1'" ;;
+        esac
+        shift
+    done
+
+    if [[ -n "$PROJECT" ]]; then
+        local subdir="$(cd "${CHIP_ROOT}/${PROJECT}" 2>/dev/null && pwd)"
+        [[ -n "$subdir" && -r "${subdir}/.gn" ]] || fail "Invalid project '${PROJECT}'"
+        PROJECT="${subdir#${CHIP_ROOT}/}"
+        [[ "$subdir" == "${CHIP_ROOT}/${PROJECT}" ]] || fail "Unable to determine project path"
+    fi
+
+    check_binary gn GN
+    check_binary ninja NINJA
+    if ! activate_build_env; then
+        check_python
+        configure_python_env
+        if ! check_binary zap-cli; then
+            download_zap
+        fi
+        finalize_build_env
+    fi
+
+    if [[ -z "$PROJECT" ]]; then
+        info "Build environment created. (Specify --project=DIR to configure a build.)"
+        return
+    fi
+
+    gn_generate "$@"
+    create_ninja_wrapper
+    info "You can now run ./ninja"
+}
+
+function gn_generate() { # [project options]
+    mkdir -p "${BUILD_ROOT}"
+    ensure_no_clobber "${BUILD_ROOT}/args.gn"
+    (
+        cd "${CHIP_ROOT}/${PROJECT}" # --root= doesn't work for gn args!
+
+        # Run gn gen with an empty args.gn first so we can list all arguments
+        info "Configuring gn build arguments"
+        echo "# ${CONFIGURE_MARKER}" >"${BUILD_ROOT}/args.gn"
+        gn -q gen "$BUILD_ROOT"
+        gn args "$BUILD_ROOT" --list --json | process_args "$@" >>"${BUILD_ROOT}/args.gn"
+
+        # Now gn gen with the arguments we have configured.
+        info "Running gn gen to generate ninja files"
+        gn -q gen "$BUILD_ROOT"
+    )
+}
+
+function process_args() { # [project options]
+    "$PYTHON" - "$@" 3<&0 <<'END'
+import collections, json, os, re, sys
```

Comment:
could we have this as a stand-alone script instead (better syntax highlighting, maybe even testing by running it as stand-alone)

Reply:
Hm, I would prefer to leave it inline since it's really just an implementation detail. I've just been testing it end to end with the configure script so far... If it becomes necessary to add some automated tests it would be relatively straightforward to test the mapping from command line args to the resulting args.gn file that way.

Reply:
I my opinion this is not easy to maintain. I can see some desire of "a single shell script" however if we need python anyway, why are we using a shell script?

Why don't we make build_examples.py be able to create light-weight builds? I believe its dependencies are also somewhat small (maybe click only). Why yet another script especially one that inlines python?

Reply:
I've pulled out the python code into `configure.utils.py` now. Python is currently needed for the build itself, so I'm opportunistically using it here for doing the JSON parsing and associated parameter mangling.

The goal of the configure script is to be simple and self-contained and to do the bare minimum necessary to set up a gn / ninja build of our C++ code within the context of an external build system. It's not intended to have any further features added to it.

The build_examples.py is a much more complex beast (5000+ lines at a cursory look) and seems to be more like a top-level meta-meta build system with specific handling for every explicitly supported target platform etc, so the goals and scope are quite different.